docs/keycloak.md

```diff
@@ -0,0 +1,356 @@
+like those logging into the Portal's frontend. Being public, it is not configured to
+give control over client scopes and roles.
+
+![Keycloak clients](rsrc/keycloak/keycloak_clients.png)
+
+The resource-server client is confidential, meaning that its client-secret
+is needed in order to authenticate against it. The portal's backend uses this client
+to authorise actions, like creating or deleting resources through the APIs.
+This is done by creating a list of roles for the resource-server client, which are
+then assigned to groups the users can belong to, which in term define the actions
+the users are allowed to perform.
+
+To perform actions that involve altering Keycloak Users and Groups from
+the Keycloak API, the Portal does have access to the default admin client within
+a Keycloak realm (by default named admin-cli).
+
+## Client Scopes
+
+The following roles are created in the resource-server client, and are assigned
+as optional client scopes:
+
+- entity:read
+- entity:write
+- entity:create
+- entity:delete
+- subscription:read
+- subscription:write
+- subscription:create
+- subscription:delete
+- device:read
+- device:write
+- device:create
+- device:delete
+- devicegroup:read
+- devicegroup:write
+- devicegroup:create
+- devicegroup:delete
+- endpoint:read
+- endpoint:write
+- endpoint:create
+- endpoint:delete
+- group:read
+- group:write
+- group:create
+- group:delete
+- user:read
+- user:write
+- user:create
+- user:delete
+
+These represent both the resource being acted upon, and the action being
+taken: reading, writing, creating and deleting. In order for a user to be authorised
+to perform a certain action, they need to belong to a Keycloak Group that has the
+corresponding role, where said Group is subgroup of the relevant Tenant.
+
+The default client scopes roles and tenants are also assigned to both resource_server
+and the portal client, the latter containing script mappers described in the
+Keycloak Mappers section below.
+
+## Tenants and Groups in Keycloak
+
+Tenants in Keycloak are represented as Groups. Subgroups of these Tenant Groups
+represent both the Service Paths used by the Orchestra Cities APIs, and the
+larger "roles" that users can have within Tenants (Admin, Developer, User, Device).
+
+![Keycloak and Portal diagram](rsrc/keycloak/tenant_groups.png)
+
+Tenant Groups are marked are such with the attributes "tenant" and "service"
+within Keycloak, whilst Service Path groups are marked with the attribute "servicePath".
+
+## Keycloak mappers
+
+![Keycloak mappers](rsrc/keycloak/keycloak_mappers.png)
+
+In order to integrate other Orchestra Cities services, such as the Portal, a series
+of Script Mappers were created as part of Client Scopes. This was done in order
+to include in the tokens given out by Keycloak the necessary information on
+Services/Tenants and Service Paths.
+
+The client scope **Tenant** was created with the following scripts:
+
```

Comment:
is not the scope that is created with this script, the content associated with user with that scope is created by the script